src/wg/util/util.d

```diff
@@ -12,6 +12,84 @@ inout(char)[] asDString(inout(char)* cstr) pure nothrow @nogc @trusted
     return cstr[0 .. len];
 }
 
+///
+enum ErrorCode : ubyte
+{
+    ///
+    success = 0,
+    ///
+    failure = 1
+}
+
+/// Template struct representing function result for functions that may return error
+struct Result(T, EC = ErrorCode) if (is(typeof(EC.success)))
+{
+    ///
+    this()(auto ref T value)
+    {
+        import core.lifetime;
+        this.value = forward!value;
+    }
+    ///
+    this(EC error, string message = "Failed")
+    {
+        assert(error != EC.success);
+        this.error = error;
+        this.message = message;
+    }
+
+    ///
+    ref inout(T) unwrap() inout pure @safe
+    {
+        import core.lifetime : move;
+        import core.exception;
+
+        if (error != EC.success)
+            throw new Exception(message);
```

Comment:
Because of this, I suggest we store line, offset and module and then set them when we throw.
So the exception is more accurate as to the source.

Reply:
Oh dear... it just gets worse and worse! >_<

Reply:
I guess that result is already huge... the horse has bolted.

Reply:
lol I agree, but may as well.